docs/concepts.rst

```diff
@@ -0,0 +1,470 @@
+
+:Removal:
+
+    An operation which removes some dataset, a sample, or some metadata which
+    existed in the parent of the commit under consideration. (Note: removing a
+    dataset also removes all samples contained in it)
+
+:Mutation:
+
+    An operation which sets: data to a sample, the value of some metadata key,
+    or a dataset schema, to a different value than what it had previously been
+    created with (Note: a dataset schema mutation is observed when a dataset is
+    removed, and a new dataset with the same name is created with a different
+    dtype/shape, all in the same commit)
+
+Merging Changes
+^^^^^^^^^^^^^^^
+
+Merging diffs solely consisting of additions and removals between branches is
+trivial, and performs exactly as one would expect from a text diff. Where
+things diverge from text is when we consider how we will merge diffs containing
+mutations.
+
+Say we have some sample in commit A, a branch is created, the sample is updated,
+and commit C is created. At the same time, someone else checks out branch whose
+HEAD is at commit A, and commits a change to the sample as well. If these
+changes are identical, they are compatible, but what if they are not? In the
+following example, we diff and merge each element of the sample array like we
+would text:
+
+::
+
+                                                   Merge ??
+      commit A          commit B            Does combining mean anything?
+
+    [[0, 1, 2],        [[0, 1, 2],               [[1, 1, 1],
+     [0, 1, 2], ----->  [2, 2, 2], ------------>  [2, 2, 2],
+     [0, 1, 2]]         [3, 3, 3]]      /         [3, 3, 3]]
+          \                            /
+           \            commit C      /
+            \                        /
+             \          [[1, 1, 1], /
+              ------->   [0, 1, 2],
+                         [0, 1, 2]]
+
+We see that a result can be generated, and can agree if this was a piece of
+text, the result would be correct. Don't be fooled, this is an abomination and
+utterly wrong/meaningless. Remember we said earlier ``"the result of a merge
+should conform to the intentions of each author"``. This merge result conforms to
+neither author's intention. The value of an array element is not isolated, every
+value affects how the entire sample is understood. The values at Commit B or
+commit C may be fine on their own, but if two samples are mutated independently
+with non-identical updates, it is a conflict that needs to be handled by the
+authors.
+
+This is the actual behavior of Hangar.
+
+::
+
+      commit A          commit B
+
+    [[0, 1, 2],        [[0, 1, 2],
+     [0, 1, 2], ----->  [2, 2, 2], ----- MERGE CONFLICT
+     [0, 1, 2]]         [3, 3, 3]]      /
+          \                            /
+           \            commit C      /
+            \                        /
+             \          [[1, 1, 1], /
+              ------->   [0, 1, 2],
+                         [0, 1, 2]]
+
```

Comment:
Needs a bit of wrap-up saying "so the merge commit author must pick one, or provide a new array" or w/e.

Reply:
Also maybe a quick word about what happens with metadata and merges (I assume the union of added keys gets added, the union of removed keys gets removed, and conflicts are generated on anything touching the same key in different ways or with different values?).